src/aws_encryption_sdk/materials_managers/__init__.py

```diff
@@ -87,21 +311,101 @@ class DecryptionMaterialsRequest(object):
+    algorithm = attr.ib(validator=instance_of(Algorithm))
+    encrypted_data_keys = attr.ib(validator=deep_iterable(member_validator=instance_of(EncryptedDataKey)))
+    encryption_context = attr.ib(
+        validator=deep_mapping(
+            key_validator=instance_of(six.string_types), value_validator=instance_of(six.string_types)
+        )
+    )
 
 
-@attr.s(hash=False)
-class DecryptionMaterials(object):
+_DEFAULT_SENTINEL = object()
+
+
+@attr.s(hash=False, init=False)
+class DecryptionMaterials(CryptographicMaterials):
     """Decryption materials returned by a crypto material manager's `decrypt_materials` method.
 
     .. versionadded:: 1.3.0
 
-    :param data_key: Plaintext data key to use with message decryption
-    :type data_key: aws_encryption_sdk.structures.DataKey
-    :param bytes verification_key: Raw signature verification key
+    .. versionadded:: 1.5.0
+
+        The **algorithm**, **data_encryption_key**, **encryption_context**, and **keyring_trace** parameters.
+
+    .. versionadded:: 1.5.0
+
+        All parameters are now optional.
+
+    :param Algorithm algorithm: Algorithm to use for encrypting message (optional)
+    :param DataKey data_encryption_key: Plaintext data key to use for encrypting message (optional)
+    :param dict encryption_context: Encryption context tied to `encrypted_data_keys` (optional)
+    :param bytes verification_key: Raw signature verification key (optional)
+    :param keyring_trace: Any KeyRing trace entries (optional)
+    :type keyring_trace: list of :class:`KeyringTrace`
     """
 
-    data_key = attr.ib(validator=attr.validators.instance_of(DataKey))
-    verification_key = attr.ib(default=None, validator=attr.validators.optional(attr.validators.instance_of(bytes)))
+    verification_key = attr.ib(default=None, repr=False, validator=optional(instance_of(bytes)))
+
+    def __init__(
+        self, data_key=_DEFAULT_SENTINEL, verification_key=None, **kwargs
+    ):  # noqa we define this in the class docstring
+
+        legacy_data_key_set = data_key is not _DEFAULT_SENTINEL
+        data_encryption_key_set = "data_encryption_key" in kwargs
+
+        if legacy_data_key_set and data_encryption_key_set:
+            raise TypeError("Either data_key or data_encryption_key can be used but not both")
+
+        if legacy_data_key_set and not data_encryption_key_set:
+            kwargs["data_encryption_key"] = data_key
+
+        for legacy_missing in ("algorithm", "encryption_context"):
+            if legacy_missing not in kwargs:
+                kwargs[legacy_missing] = None
+
+        super(DecryptionMaterials, self).__init__(**kwargs)
+
+        self._setattr("verification_key", verification_key)
+        attr.validate(self)
+
+    @property
+    def data_key(self):
+        """Backwards-compatible shim for access to data key."""
+        return self.data_encryption_key
+
+    def add_data_encryption_key(self, data_encryption_key, keyring_trace):
```

Comment:
Don't we need a `_validate_data_encryption_key` call here?

Reply:
That's strictly a shim for access via the old API. We put guardrails around changing the data encryption key and we check it when you set it through those guardrails. I don't think there's really much benefit checking it *again* in this one specific access path.

If we want to validate the data encryption key on access, we should do that via a `property` pattern for `CryptographicMaterials.data_encryption_key`.